autoconnect/autoconnect-ws/src/ping.rs

```diff
@@ -0,0 +1,109 @@
+use tokio::time::{interval, Interval};
+
+use autoconnect_common::{broadcast::Broadcast, protocol::ServerMessage};
+use autoconnect_settings::Settings;
+use autoconnect_ws_sm::WebPushClient;
+
+use crate::{error::WSError, session::Session};
+
+#[derive(Debug)]
+enum Waiting {
+    /// Waiting to send a WebSocket Ping (or WebPush Broadcast) to the Client
+    ToPing,
+    /// Waiting for the Client to respond to our WebSocket Ping with a Pong
+    ForPong,
+}
+
+/// Manages WebSocket Pings sent to the Client
+///
+/// We automatically send WebSocket Pings (or WebPush Broadcasts, if one is
+/// pending) every `auto_ping_interval`. If the Client fails to respond to the
+/// Ping with a Pong within the `auto_ping_timeout` interval we drop their
+/// connection
+#[derive(Debug)]
+pub struct PingManager {
+    /// Waiting to Ping or timeout recieving a Pong
+    waiting: Waiting,
+    ping_or_timeout: Interval,
+}
+
+impl PingManager {
+    pub async fn new(settings: &Settings) -> PingManager {
+        let mut ping = interval(settings.auto_ping_interval);
```

Comment:
It feels a bit confusing to thave basically two different `ticks()` here.
I understand what they are, but a casual / fast read kind of blurs them.

I wonder if it might be useful to call this one `ping_clock`, since it's an `Interval`?

Reply:
Is this any clearer?